### PR TITLE
Support Pacman package manager

### LINKED ISSUE #57: Support querying Pacman package manager
The package resource could easily add support for the `pacman` package manager used in Arch Linux. If we add support, I'll switch distributive's package check to use goss's code. Would you merge a PR to this effect?

### PR DESCRIPTION
**Pull request by siddharthist**
Fixes #57 

Tests added and passing. Thanks for reviewing!

### CODE REVIEW

**Comment by aelsabbahy** [comment]
Tests aren't running for arch, add them integration-test task: https://github.com/aelsabbahy/goss/pull/60/files#diff-b67911656ef5d18c4ae36cb6741b7965R43

**Comment by siddharthist** [comment]
I'm not convinced that the error is related. If you search "exit 1" in the full log, the error is one of service detection.

**Comment by aelsabbahy** [comment]
The travis results are hard to read because they're run in parallel, just run `make arch` locally and you'll see the errors.

It might be that `service` is not handling arch correctly, I haven't had time to dig into it to understand why.

**Comment by siddharthist** [comment]
@aelsabbahy It looks to me like that's the case: http://pastebin.com/raw/nnx4fh8f

**Comment by aelsabbahy** [comment]
It's part of goss-shared.json. `service` is returning running: true in arch for a non existent service.

**Comment by siddharthist** [comment]
@aelsabbahy What would you like to do about that? It doesn't seem like a problem related to this PR.

**Comment by aelsabbahy** [comment]
Don't think we should have half-baked arch-linux support. It would be very easy for someone to assume that Arch is supported when it's not fully functional.

Feel free to create another PR for service or bundle the two up.

Take a look at #35 for reference, this was a PR to add alpine Linux support from another contributor.

**Comment by siddharthist** [comment]
@aelsabbahy This doesn't strike me as an Arch Linux specific problem, Arch uses systemd for service management just like many other distributions.

**Comment by aelsabbahy** [comment]
If you have an Archlinux virtual or physical machine, I would test it there. If it works, then it might be the Dockerfile or docker run calls that need to be fixed to support systemd properly.

**Comment by siddharthist** [comment]
@aelsabbahy Good thinking! Looks like it works properly on my machine, I'll try to get a better Dockerfile.
```
$ mv integration-tests/goss/goss-shared.json goss.yaml
$ goss validate
...................

Total Duration: 0.502s
Count: 19, Failed: 0

```

**Comment by siddharthist** [comment]
I think the problem is likely that [goss ignores errors when querying services](https://github.com/aelsabbahy/goss/blob/d28f3cc6d708fb012ea614acf712eb56712a7de3/resource/service.go#L35-L36). If this were not the case, we'd likely see an error here, instead of a false positive.

**Comment by siddharthist** [comment]
It turns out that `dbus` is not running in the Arch docker container, which makes a lot of sense (there's no need for services in a single-process container). I can't find an easy way to get it running, either. I think we should just remove the service check from the shared goss tests, what do you think @aelsabbahy?

**Comment by aelsabbahy** [comment]
#64 Fixes the silent error when system.go is misconfigured to init for a system that wasn't using init, which is the case here. Rebasing on master will show you a clearer error message.

system.go needs to be changed to mark archlinux as NewServiceDbus rather than using the default NewServiceInit.

I would look into doing something like: https://github.com/maci0/docker-systemd-unpriv for running systemd inside docker.

**Comment by siddharthist** [comment]
@aelsabbahy I fixed the error message problem more broadly in 2f668ad and added arch with NewServiceDbus in ce31e45. I'll take a look at that example and let you know if I can come up with anything!

**Comment by siddharthist** [comment]
@aelsabbahy Would you be willing to help get DBus running in Docker? I'm not sure I know how.

**Comment by aelsabbahy** [comment]
I don't mind looking into this, but my schedule has been pretty busy, so it might take me a while to get around to it.

FYI, 2f668ad addresses the goss `add` and `autoadd` case, I don't believe that's used when doing a `validate`

**Comment by siddharthist** [comment]
@aelsabbahy Thanks for all your feedback, it's much appreciated! I got dbus running in docker, but it requires some mounts on the host (and the host has to be running systemd). This conforms to the example you provided and runs just fine locally, but doesn't seem to like to run in Travis. What would you like to do? 

One option would be to skip `goss-shared.json` or just the service `foobar` check when `$os == arch` and `$CI == true`.

**Comment by aelsabbahy** [comment]
Wasn't able to get a successful run locally, can you change:
`make -j10`
to:
`make`

In the travis.yml file https://github.com/aelsabbahy/goss/blob/master/.travis.yml#L25

This will make it easier to see why it's failing in travis.

**Comment by siddharthist** [comment]
@aelsabbahy Are you testing locally on a distro running systemd?

**Comment by siddharthist** [comment]
Here's the error: 
```
Failures:

foobar: enabled: Error: No such interface '\''org.freedesktop.DBus.Properties'\'' on object at path /org/freedesktop/systemd1/unit/foobar_2eservice
foobar: running: Error: No such interface '\''org.freedesktop.DBus.Properties'\'' on object at path /org/freedesktop/systemd1/unit/foobar_2eservice
```

**Comment by aelsabbahy** [comment]
I was able to get it working with centos7 using the instructions from: https://hub.docker.com/_/centos/

I think that rules out travis, I wonder if there's a way to get systemd in Arch to behave the same way as it does in redhat/centOS.

I would try to see if you can get it working in a similar fashion as centos7 one more time.. If you can't I guess we can move the service test out of shared, and into a 2nd shared file that's imported by all the other integration tests.

Here's centos7 succeeding and arch failing for reference:
* https://travis-ci.org/aelsabbahy/goss/builds/120183315
* 8a3bf5f

**Comment by siddharthist** [comment]
@aelsabbahy How would I remove it from goss-shared.json and still have it included in testing? Sorry, I'm trying to read through the testing scripts, but I'm having a hard time understanding them.

**Comment by aelsabbahy** [comment]
I would create a 2nd shared json file with the service entry in it and import that in all the relevant goss.json test fixtures. Example: https://github.com/aelsabbahy/goss/blob/master/integration-tests/goss/centos6/goss.json#L66

**Comment by siddharthist** [comment]
@aelsabbahy Oh, I see. Thank you!

**Comment by siddharthist** [comment]
@aelsabbahy Finally got passing tests :smile:

File: system/package_pacman.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	p.loaded = true
+	// TODO: extract versions
+	cmd := util.NewCommand("pacman", "-Q", "--color", "never", "--noconfirm", p.name)
+	if err := cmd.Run(); err != nil {
+		return
+	}
+	p.installed = true
+	// the output format is "pkgname version\n", so if we split the string on
+	// whitespace, the version is the second item.
+	p.versions = []string{strings.Fields(cmd.Stdout.String())[1]}
```
Comment (aelsabbahy) [inline system/package_pacman.go]: Does pacman support multiple versions of the same package being installed? (ex: multiple kernels)

This code looks like it assumes only 1 package is installed.

Comment (siddharthist) [inline system/package_pacman.go]: I'm not aware of any support for multiple package versions.

File: integration-tests/test.sh [hunk 2]
```diff
@@ -25,17 +29,19 @@ for arch in amd64 386;do
   out=$(docker exec goss_int_test_$os bash -c "time /tmp/goss/$os/goss-linux-$arch -g /tmp/goss/$os/goss.json validate")
   echo "$out"
 
-  grep -q 'Count: 39, Failed: 0' <<<"$out"
+  egrep -q 'Count: [0-9]{2}, Failed: 0' <<<"$out"
```
Comment (aelsabbahy) [inline integration-tests/test.sh]: The original code is brittle, but until unit tests are in place for parsing JSON/YAML, it's the only validation there is that all the tests ran and none were skipped. Can you undo this change.

Comment (siddharthist) [inline integration-tests/test.sh]: What do you want me to do? Just add a conditional for if `$os == arch`? It's a different number of tests.

Comment (aelsabbahy) [inline integration-tests/test.sh]: oh crap, forgot arch is unique.. screw it.. keep it as is, I'll fix later.

**Comment by aelsabbahy** [comment]
This looks good enough to merge, can you squash the PR and I'll merge it when tests pass again.

**Comment by siddharthist** [comment]
@aelsabbahy Github has a new squash-on-merge button you can use :smile:

**Comment by aelsabbahy** [comment]
Today I learned :)

Thanks for the contribution and thanks for the squash-on-merge tip!

**Comment by aelsabbahy** [comment]
FYI.. not sure if random travis-ci hiccup or related to squash-on-merge button, but a build didn't kick off for master *shrug*.

**Comment by siddharthist** [comment]
@aelsabbahy Thanks for merging :fireworks:
